File: t075/ptr.c
/*
 * Pointer assignment demo 
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct str_t {
	int x, y;
};

typedef struct {
	int id;
	int *init;

	char buf_arr[32];
	char *buf_ptr;

	struct str_t s1;
	struct str_t *s2;
} request_t;

void pprint_req(char *name, request_t *req) {
	printf("\nRequest %s:\n", name);
	printf("\taddress: %p\n", (void *)req);

	printf("\tid: %d\n", req->id);
	printf("\tinit: %p\n", (void *)req->init);
	printf("\t  *init: %d\n", *req->init);

	printf("\tbuf_arr: %p\n", (void *)req->buf_arr);
	printf("\t  *buff_arr: \"%s\"\n", req->buf_arr);

	printf("\tbuf_ptr: %p\n", (void *)req->buf_ptr);
	printf("\t  *buf_ptr: \"%s\"\n", req->buf_ptr);

	printf("\ts1 (str_t):\n");
	printf("\t  req->s1.x: %d\n", req->s1.x);
	printf("\t  req->s1.y: %d\n", req->s1.y);

	printf("\ts2 (str_t *):\n");
	printf("\t  &s2: %p\n", (void *)req->s2);
	if (req->s2 != NULL) {
		printf("\t  req->s2->x: %d\n", req->s2->x);
		printf("\t  req->s2->y: %d\n", req->s2->y);
	}

	printf("\n");
	fflush(stdout);
}

int main(void) {
	request_t *a = malloc(sizeof(request_t));
	memset(a, 0, sizeof(request_t));
	strcpy(a->buf_arr, "Yo yo yo, world!");
	a->id = 42;
	a->init = malloc(sizeof(int));
	*a->init = 505;
	a->s1.x = a->s1.y = 17;
	pprint_req("a", a);

	request_t *b = a;
	pprint_req("b (b=a)", b);

	request_t *c = malloc(sizeof(request_t));
	memset(c, 0, sizeof(request_t));
	*c = *a;
	pprint_req("c (*c=*a)", c);

	printf("-------------------------------------------------------------\n");
	printf("Modifying a->buf_arr content...\n");
	strcpy(a->buf_arr, "Modified content!");
	pprint_req("a", a);
	pprint_req("b (b=a)", b);
	pprint_req("c (*c=*a)", c);

	printf("-------------------------------------------------------------\n");
	printf("Allocating a->buf_ptr content...\n");
	a->buf_ptr = malloc(32);
	strcpy(a->buf_ptr, "New a->buf_ptr content!");
	pprint_req("a", a);
	pprint_req("b (b=a)", b);
	pprint_req("c (*c=*a)", c);

	printf("-------------------------------------------------------------\n");
	printf("Allocating a->s2 content...\n");
	a->s2 = malloc(sizeof(struct str_t));
	a->s2->x = 1;
	a->s2->y = 2;
	pprint_req("a", a);
	pprint_req("b (b=a)", b);
	pprint_req("c (*c=*a)", c);

	printf("-------------------------------------------------------------\n");
	printf("Modifying *a->init content...\n");
	*a->init = 333;
	pprint_req("a", a);
	pprint_req("b (b=a)", b);
	pprint_req("c (*c=*a)", c);

	return 0;
}

File: t075/ptr.go
/*
 * Pointer assignment demo
 */
package main

import "fmt"

type str_t struct {
	x, y int
}

type Request struct {
	id   int
	init *int

	buf string

	s1 str_t
	s2 *str_t
}

func (req *Request) PrettyPrintReq(name string) {
	fmt.Printf("\nRequest %s:\n", name)
	fmt.Printf("\taddress: %p\n", req)

	fmt.Printf("\tid: %d\n", req.id)
	fmt.Printf("\tinit: %p\n", req.init)
	fmt.Printf("\t  *init: %d\n", *req.init)

	fmt.Printf("\tbuf:\n")
	fmt.Printf("\t  &buf: %p\n", &req.buf)
	fmt.Printf("\t  buf: \"%s\"\n", req.buf)

	fmt.Printf("\ts1 (str_t):\n")
	fmt.Printf("\t  req.s1.x: %d\n", req.s1.x)
	fmt.Printf("\t  req.s1.y: %d\n", req.s1.y)

	fmt.Printf("\ts2 (*str_t):\n")
	fmt.Printf("\t  s2: %p\n", req.s2)
	if req.s2 != nil {
		fmt.Printf("\t  req.s2.x: %d\n", req.s2.x)
		fmt.Printf("\t  req.s2.y: %d\n", req.s2.y)
	}

	fmt.Println()
}

func main() {
	init_tmp := new(int)
	*init_tmp = 505
	a := &Request{
		id:   42,
		init: init_tmp,
		buf:  "Yo yo yo, world!",
		s1:   str_t{17, 17},
	}
	a.PrettyPrintReq("a")

	b := a
	b.PrettyPrintReq("b (b=a)")

	c := new(Request)
	*c = *a
	c.PrettyPrintReq("c (*c=*a)")

	fmt.Printf("------------------------------------------------------\n")
	fmt.Printf("Modifying a.buf content...\n")
	a.buf = "Modified content!"
	a.PrettyPrintReq("a")
	b.PrettyPrintReq("b (b=a)")
	c.PrettyPrintReq("c (*c=*a)")

	fmt.Printf("------------------------------------------------------\n")
	fmt.Printf("Allocating a.s2 content...\n")
	a.s2 = &str_t{1, 2}
	a.PrettyPrintReq("a")
	b.PrettyPrintReq("b (b=a)")
	c.PrettyPrintReq("c (*c=*a)")

	fmt.Printf("------------------------------------------------------\n")
	fmt.Printf("Modifying *a.init...\n")
	*a.init = 333
	a.PrettyPrintReq("a")
	b.PrettyPrintReq("b (b=a)")
	c.PrettyPrintReq("c (*c=*a)")

	return
}
